From 5afdcb33d0f06b52fd544ef1e0cb7c80783a2ef6 Mon Sep 17 00:00:00 2001
From: Carlo Monjaraz <carlomonjaraztec@proton.me>
Date: Thu, 20 Aug 2026 02:52:36 +0200
Subject: [PATCH 1/4] Add native VTK XML Unstructured Grid (.vtu) and ParaView
 (.pvd) exporter

- Implement write2vtu.c with support for all 1D/2D/3D linear & quadratic elements
- Add point data fields: Displacements (vector), Stress (tensor), von Mises, and Principal Stresses (P1, P2, P3)
- Add multi-step and frequency mode export support with .pvd collection timeline
- Minimally invasive integration into sendSet.c and Makefile.inc
---
 cgx_2.23/src/Makefile.inc   |   1 +
 cgx_2.23/src/cgx.h          |   1 +
 cgx_2.23/src/setFunktions.c |  11 +
 cgx_2.23/src/write2vtu.c    | 481 ++++++++++++++++++++++++++++++++++++
 4 files changed, 494 insertions(+)
 create mode 100644 cgx_2.23/src/write2vtu.c

diff --git a/cgx_2.23/src/Makefile.inc b/cgx_2.23/src/Makefile.inc
index ac3e648..c8ad606 100644
--- a/cgx_2.23/src/Makefile.inc
+++ b/cgx_2.23/src/Makefile.inc
@@ -117,6 +117,7 @@ SLIB = \
  write2duns.c \
  write2frd.c \
  write2foam.c \
+ write2vtu.c \
  write2nas.c \
  write2samcef.c \
  write2tochnog.c \
diff --git a/cgx_2.23/src/cgx.h b/cgx_2.23/src/cgx.h
index 6b4cb18..c5b3aeb 100755
--- a/cgx_2.23/src/cgx.h
+++ b/cgx_2.23/src/cgx.h
@@ -407,6 +407,7 @@ void splitBiasDiv(int *ptrdiv, double *ptrbias);
 int improveBadTr3(char *setname);
 void printHash();
 int write2foam(char *setname, int strings, char **string, Summen *anz, Nodes *node, Faces *face, Elements *e_enqire, Sets *set, Datasets *lcase );
+int write2vtu(char *setname, int strings, char **string, Summen *anz, Nodes *node, Faces *face, Elements *e_enqire, Sets *set, Datasets *lcase );
 int write2dolfyn(char *setname, int strings, char **string, Summen *anz, Nodes *node, Faces *face, Elements *e_enqire, Sets *set);
 int readFoam(char *datin, Summen *anz, Sets **sptr, Nodes **nptr, Elements **eptr, Datasets **lptr );
 int readVtk(char *datin, Summen *anz, Nodes **nptr, Elements **eptr, Datasets **lptr );
diff --git a/cgx_2.23/src/setFunktions.c b/cgx_2.23/src/setFunktions.c
index 37c7587..ec2f89f 100644
--- a/cgx_2.23/src/setFunktions.c
+++ b/cgx_2.23/src/setFunktions.c
@@ -23729,6 +23729,17 @@ void pre_write( char *record )
     /* free dat */
     for(i=0; i<length; i++) free(dat[i]); free(dat);
   }
+  else if (compare( format, "vtu", 3)== 3 || compare( format, "vtk", 3)== 3)
+  {
+    length= strsplt( record, ' ', &dat);
+    descalNodes ( anz->n, node, scale);
+    if(length>2) write2vtu( setname, length-2, &dat[2], anz, node, face, e_enqire, set, lcase );
+    else write2vtu( setname, 0, NULL, anz, node, face, e_enqire, set, lcase );
+    scalNodes ( anz->n, node, scale);
+
+    /* free dat */
+    for(i=0; i<length; i++) free(dat[i]); free(dat);
+  }
   else if (compare( format, "isaac", 3)== 3)
   {
     length= strsplt( record, ' ', &dat);
diff --git a/cgx_2.23/src/write2vtu.c b/cgx_2.23/src/write2vtu.c
new file mode 100644
index 0000000..73055b0
--- /dev/null
+++ b/cgx_2.23/src/write2vtu.c
@@ -0,0 +1,481 @@
+/* --------------------------------------------------------------------  */
+/*                          CALCULIX                                     */
+/*                   - GRAPHICAL INTERFACE -                             */
+/*                                                                       */
+/*     A 3-dimensional pre- and post-processor for finite elements       */
+/*              Copyright (C) 1996 Klaus Wittig                          */
+/*                                                                       */
+/*     This program is free software; you can redistribute it and/or     */
+/*     modify it under the terms of the GNU General Public License as    */
+/*     published by the Free Software Foundation; version 2 of           */
+/*     the License.                                                      */
+/* --------------------------------------------------------------------  */
+
+/* write2vtu.c - Export mesh and field datasets to VTK XML Unstructured Grid (.vtu) */
+/* and ParaView Data collection (.pvd)                                               */
+
+#include <cgx.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+
+#ifndef M_PI
+#define M_PI 3.14159265358979323846
+#endif
+
+extern int cur_lc;
+int readfrdblock( int lc, Summen *anz, Nodes *node, Datasets *lcase );
+void calcDatasets( int num_olc, Summen *anz, Nodes *node, Datasets *lcase );
+
+/* Calculate Principal Stresses P1 >= P2 >= P3 and von Mises stress analytically */
+static void calc_principal_and_mises(double sxx, double syy, double szz,
+                                     double sxy, double syz, double szx,
+                                     double *p1, double *p2, double *p3, double *vm)
+{
+  double p, s11, s22, s33, J2, J3, q, arg, phi;
+  double r1, r2, r3, tmp;
+
+  /* Mean hydrostatic stress */
+  p = (sxx + syy + szz) / 3.0;
+
+  /* Deviatoric stress components */
+  s11 = sxx - p;
+  s22 = syy - p;
+  s33 = szz - p;
+
+  /* Second invariant J2 */
+  J2 = 0.5 * (s11 * s11 + s22 * s22 + s33 * s33) + sxy * sxy + syz * syz + szx * szx;
+
+  /* Von Mises equivalent stress: sqrt(3 * J2) */
+  if (J2 > 0.0) *vm = sqrt(3.0 * J2);
+  else *vm = 0.0;
+
+  /* Determinant of deviatoric stress (Third invariant J3) */
+  J3 = s11 * (s22 * s33 - syz * syz)
+     - sxy * (sxy * s33 - syz * szx)
+     + szx * (sxy * syz - s22 * szx);
+
+  if (J2 < 1e-20)
+  {
+    *p1 = p;
+    *p2 = p;
+    *p3 = p;
+    return;
+  }
+
+  q = 2.0 * sqrt(J2 / 3.0);
+  arg = (3.0 * J3) / (q * J2);
+  if (arg > 1.0) arg = 1.0;
+  if (arg < -1.0) arg = -1.0;
+
+  phi = acos(arg) / 3.0;
+
+  r1 = p + q * cos(phi);
+  r2 = p + q * cos(phi + (2.0 * M_PI / 3.0));
+  r3 = p + q * cos(phi + (4.0 * M_PI / 3.0));
+
+  /* Sort descending: p1 >= p2 >= p3 */
+  if (r1 < r2) { tmp = r1; r1 = r2; r2 = tmp; }
+  if (r2 < r3) { tmp = r2; r2 = r3; r3 = tmp; }
+  if (r1 < r2) { tmp = r1; r1 = r2; r2 = tmp; }
+
+  *p1 = r1;
+  *p2 = r2;
+  *p3 = r3;
+}
+
+/* Map CGX element types to VTK cell types */
+static int get_vtk_cell_type(int cgx_type, int *num_nodes)
+{
+  switch (cgx_type)
+  {
+    case 1:  *num_nodes = 8;  return 12; /* VTK_HEXAHEDRON */
+    case 2:  *num_nodes = 6;  return 13; /* VTK_WEDGE */
+    case 3:  *num_nodes = 4;  return 10; /* VTK_TETRA */
+    case 4:  *num_nodes = 20; return 25; /* VTK_QUADRATIC_HEXAHEDRON */
+    case 5:  *num_nodes = 15; return 26; /* VTK_QUADRATIC_WEDGE */
+    case 6:  *num_nodes = 10; return 24; /* VTK_QUADRATIC_TETRA */
+    case 7:  *num_nodes = 3;  return 5;  /* VTK_TRIANGLE */
+    case 8:  *num_nodes = 6;  return 22; /* VTK_QUADRATIC_TRIANGLE */
+    case 9:  *num_nodes = 4;  return 9;  /* VTK_QUAD */
+    case 10: *num_nodes = 8;  return 23; /* VTK_QUADRATIC_QUAD */
+    case 11: *num_nodes = 2;  return 3;  /* VTK_LINE */
+    case 12: *num_nodes = 3;  return 21; /* VTK_QUADRATIC_EDGE */
+    default: *num_nodes = 0;  return 0;
+  }
+}
+
+/* Write a single .vtu file for a specific step or current dataset */
+static int write_single_vtu_file(const char *filename, int target_step,
+                                 Summen *anz, Nodes *node, Elements *elem,
+                                 Sets *set, int setNr, Datasets *lcase,
+                                 int num_pts, int *pts_nodenr, int *nodeMap,
+                                 int num_elems, int *elem_indices)
+{
+  FILE *fp;
+  int i, j, k, lc, eid, nid, vtk_type, n_nodes;
+  long offset = 0;
+  double sxx, syy, szz, sxy, syz, szx;
+  double p1, p2, p3, vm;
+
+  fp = fopen(filename, "w");
+  if (!fp)
+  {
+    printf("ERROR: Could not open output file: %s\n", filename);
+    return 0;
+  }
+
+  fprintf(fp, "<?xml version=\"1.0\"?>\n");
+  fprintf(fp, "<VTKFile type=\"UnstructuredGrid\" version=\"0.1\" byte_order=\"LittleEndian\">\n");
+  fprintf(fp, "  <UnstructuredGrid>\n");
+  fprintf(fp, "    <Piece NumberOfPoints=\"%d\" NumberOfCells=\"%d\">\n", num_pts, num_elems);
+
+  /* 1. Point Coordinates */
+  fprintf(fp, "      <Points>\n");
+  fprintf(fp, "        <DataArray type=\"Float64\" Name=\"Points\" NumberOfComponents=\"3\" format=\"ascii\">\n");
+  for (i = 0; i < num_pts; i++)
+  {
+    nid = pts_nodenr[i];
+    fprintf(fp, "          %.9e %.9e %.9e\n", node[nid].nx, node[nid].ny, node[nid].nz);
+  }
+  fprintf(fp, "        </DataArray>\n");
+  fprintf(fp, "      </Points>\n");
+
+  /* 2. Cells (Topology, Offsets, Types) */
+  fprintf(fp, "      <Cells>\n");
+
+  /* Connectivity */
+  fprintf(fp, "        <DataArray type=\"Int64\" Name=\"connectivity\" format=\"ascii\">\n");
+  for (i = 0; i < num_elems; i++)
+  {
+    eid = elem_indices[i];
+    vtk_type = get_vtk_cell_type(elem[eid].type, &n_nodes);
+    fprintf(fp, "         ");
+    for (j = 0; j < n_nodes; j++)
+    {
+      nid = elem[eid].nod[j];
+      fprintf(fp, " %d", (nid >= 0 && nid <= anz->nmax) ? nodeMap[nid] : 0);
+    }
+    fprintf(fp, "\n");
+  }
+  fprintf(fp, "        </DataArray>\n");
+
+  /* Offsets */
+  fprintf(fp, "        <DataArray type=\"Int64\" Name=\"offsets\" format=\"ascii\">\n");
+  offset = 0;
+  for (i = 0; i < num_elems; i++)
+  {
+    eid = elem_indices[i];
+    get_vtk_cell_type(elem[eid].type, &n_nodes);
+    offset += n_nodes;
+    fprintf(fp, "          %ld\n", offset);
+  }
+  fprintf(fp, "        </DataArray>\n");
+
+  /* Types */
+  fprintf(fp, "        <DataArray type=\"UInt8\" Name=\"types\" format=\"ascii\">\n");
+  for (i = 0; i < num_elems; i++)
+  {
+    eid = elem_indices[i];
+    vtk_type = get_vtk_cell_type(elem[eid].type, &n_nodes);
+    fprintf(fp, "          %d\n", vtk_type);
+  }
+  fprintf(fp, "        </DataArray>\n");
+  fprintf(fp, "      </Cells>\n");
+
+  /* 3. Point Data Fields */
+  fprintf(fp, "      <PointData>\n");
+  if (anz->l > 0 && lcase != NULL)
+  {
+    for (lc = 0; lc < anz->l; lc++)
+    {
+      /* Filter datasets by target_step if specified (target_step > 0) */
+      if (target_step > 0 && lcase[lc].step_number != target_step) continue;
+      if (target_step == 0 && lc != cur_lc && anz->l > 1)
+      {
+        if (lcase[lc].step_number != lcase[cur_lc].step_number) continue;
+      }
+
+      /* Ensure dataset is loaded in memory */
+      if (!lcase[lc].loaded)
+      {
+        readfrdblock(lc, anz, node, lcase);
+        calcDatasets(lc, anz, node, lcase);
+      }
+
+      if (!lcase[lc].loaded || lcase[lc].dat == NULL) continue;
+
+      /* Displacements Vector (3 components) */
+      if (compare(lcase[lc].name, "DISP", 4) == 4 || (lcase[lc].ncomps >= 3 && compare(lcase[lc].compName[0], "D1", 2) == 2))
+      {
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"Displacements\" NumberOfComponents=\"3\" format=\"ascii\">\n");
+        for (i = 0; i < num_pts; i++)
+        {
+          nid = pts_nodenr[i];
+          fprintf(fp, "          %.6e %.6e %.6e\n",
+                  lcase[lc].dat[0][nid], lcase[lc].dat[1][nid], lcase[lc].dat[2][nid]);
+        }
+        fprintf(fp, "        </DataArray>\n");
+      }
+      /* Stress Tensor & Precomputed Invariants */
+      else if (compare(lcase[lc].name, "STRESS", 6) == 6 || lcase[lc].ncomps >= 6)
+      {
+        /* 6-component Stress Tensor */
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"Stress\" NumberOfComponents=\"6\" format=\"ascii\">\n");
+        for (i = 0; i < num_pts; i++)
+        {
+          nid = pts_nodenr[i];
+          fprintf(fp, "          %.6e %.6e %.6e %.6e %.6e %.6e\n",
+                  lcase[lc].dat[0][nid], lcase[lc].dat[1][nid], lcase[lc].dat[2][nid],
+                  lcase[lc].dat[3][nid], lcase[lc].dat[4][nid], lcase[lc].dat[5][nid]);
+        }
+        fprintf(fp, "        </DataArray>\n");
+
+        /* Precomputed von Mises Scalar */
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"vonMises\" NumberOfComponents=\"1\" format=\"ascii\">\n");
+        for (i = 0; i < num_pts; i++)
+        {
+          nid = pts_nodenr[i];
+          sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+          sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+          calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+          fprintf(fp, "          %.6e\n", vm);
+        }
+        fprintf(fp, "        </DataArray>\n");
+
+        /* Principal Stresses */
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"PrincipalStress_1\" NumberOfComponents=\"1\" format=\"ascii\">\n");
+        for (i = 0; i < num_pts; i++)
+        {
+          nid = pts_nodenr[i];
+          sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+          sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+          calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+          fprintf(fp, "          %.6e\n", p1);
+        }
+        fprintf(fp, "        </DataArray>\n");
+
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"PrincipalStress_2\" NumberOfComponents=\"1\" format=\"ascii\">\n");
+        for (i = 0; i < num_pts; i++)
+        {
+          nid = pts_nodenr[i];
+          sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+          sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+          calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+          fprintf(fp, "          %.6e\n", p2);
+        }
+        fprintf(fp, "        </DataArray>\n");
+
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"PrincipalStress_3\" NumberOfComponents=\"1\" format=\"ascii\">\n");
+        for (i = 0; i < num_pts; i++)
+        {
+          nid = pts_nodenr[i];
+          sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+          sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+          calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+          fprintf(fp, "          %.6e\n", p3);
+        }
+        fprintf(fp, "        </DataArray>\n");
+      }
+      /* Scalar Datasets (Temperature, PEEQ, Pressure, etc.) */
+      else if (lcase[lc].ncomps == 1)
+      {
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"%s\" NumberOfComponents=\"1\" format=\"ascii\">\n", lcase[lc].name);
+        for (i = 0; i < num_pts; i++)
+        {
+          nid = pts_nodenr[i];
+          fprintf(fp, "          %.6e\n", lcase[lc].dat[0][nid]);
+        }
+        fprintf(fp, "        </DataArray>\n");
+      }
+    }
+  }
+  fprintf(fp, "      </PointData>\n");
+
+  /* 4. Cell Data (Element IDs & Materials) */
+  fprintf(fp, "      <CellData>\n");
+  fprintf(fp, "        <DataArray type=\"Int32\" Name=\"Element_ID\" format=\"ascii\">\n");
+  for (i = 0; i < num_elems; i++)
+  {
+    eid = elem_indices[i];
+    fprintf(fp, "          %d\n", elem[eid].nr);
+  }
+  fprintf(fp, "        </DataArray>\n");
+
+  fprintf(fp, "        <DataArray type=\"Int32\" Name=\"Material_ID\" format=\"ascii\">\n");
+  for (i = 0; i < num_elems; i++)
+  {
+    eid = elem_indices[i];
+    fprintf(fp, "          %d\n", elem[eid].mat);
+  }
+  fprintf(fp, "        </DataArray>\n");
+  fprintf(fp, "      </CellData>\n");
+
+  fprintf(fp, "    </Piece>\n");
+  fprintf(fp, "  </UnstructuredGrid>\n");
+  fprintf(fp, "</VTKFile>\n");
+
+  fclose(fp);
+  printf("  -> Written: %s (%d nodes, %d cells)\n", filename, num_pts, num_elems);
+  return 1;
+}
+
+/* Main entry point for send <set> vtu [all] */
+int write2vtu(char *setname, int strings, char **string, Summen *anz,
+              Nodes *node, Faces *face, Elements *elem, Sets *set, Datasets *lcase)
+{
+  int setNr, i, j, k, eid, nid, n_nodes, vtk_type;
+  int num_pts = 0, num_elems = 0;
+  int *nodeMap = NULL;
+  int *pts_nodenr = NULL;
+  int *elem_indices = NULL;
+  int export_all_steps = 0;
+  int num_steps = 0, *step_list = NULL;
+  double *step_times = NULL;
+  char vtu_filename[MAX_LINE_LENGTH];
+  char pvd_filename[MAX_LINE_LENGTH];
+  FILE *pvd_fp = NULL;
+
+  if (anz->n == 0 || anz->e == 0)
+  {
+    printf("ERROR: No mesh available to export.\n");
+    return 0;
+  }
+
+  setNr = getSetNr(setname);
+  if (setNr < 0)
+  {
+    printf("ERROR: Set %s does not exist.\n", setname);
+    return 0;
+  }
+
+  /* Check if \"all\" steps requested */
+  if (strings > 0 && string != NULL && string[0] != NULL)
+  {
+    if (compare(string[0], "all", 3) == 3)
+    {
+      export_all_steps = 1;
+    }
+  }
+
+  printf("\n=== Exporting to VTK XML Unstructured Grid (.vtu) ===\n");
+  printf("Set: %s (mode: %s)\n", setname, export_all_steps ? "all steps (.pvd)" : "active step (.vtu)");
+
+  /* 1. Identify Elements in Set */
+  num_elems = set[setNr].anz_e;
+  if (num_elems <= 0)
+  {
+    printf("ERROR: Set %s contains no elements.\n", setname);
+    return 0;
+  }
+  elem_indices = set[setNr].elem;
+
+  /* 2. Map Unique Nodes Used in Set */
+  nodeMap = (int *)malloc((anz->nmax + 1) * sizeof(int));
+  for (i = 0; i <= anz->nmax; i++) nodeMap[i] = -1;
+
+  /* Mark all nodes referenced by elements */
+  for (i = 0; i < num_elems; i++)
+  {
+    eid = elem_indices[i];
+    vtk_type = get_vtk_cell_type(elem[eid].type, &n_nodes);
+    if (vtk_type == 0) continue;
+    for (j = 0; j < n_nodes; j++)
+    {
+      nid = elem[eid].nod[j];
+      if (nid > 0 && nid <= anz->nmax && nodeMap[nid] == -1)
+      {
+        nodeMap[nid] = num_pts++;
+      }
+    }
+  }
+
+  if (num_pts == 0)
+  {
+    printf("ERROR: No valid points found in set %s.\n", setname);
+    free(nodeMap);
+    return 0;
+  }
+
+  /* Create array of user node IDs indexed by local 0-based point ID */
+  pts_nodenr = (int *)malloc(num_pts * sizeof(int));
+  for (i = 0; i <= anz->nmax; i++)
+  {
+    if (nodeMap[i] >= 0 && nodeMap[i] < num_pts)
+    {
+      pts_nodenr[nodeMap[i]] = i;
+    }
+  }
+
+  /* 3. Export Data */
+  if (export_all_steps && anz->l > 0)
+  {
+    /* Find unique steps */
+    step_list = (int *)malloc(anz->l * sizeof(int));
+    step_times = (double *)malloc(anz->l * sizeof(double));
+    for (i = 0; i < anz->l; i++)
+    {
+      int step = lcase[i].step_number;
+      int exists = 0;
+      for (j = 0; j < num_steps; j++)
+      {
+        if (step_list[j] == step) { exists = 1; break; }
+      }
+      if (!exists)
+      {
+        step_list[num_steps] = step;
+        step_times[num_steps] = (double)lcase[i].value;
+        num_steps++;
+      }
+    }
+
+    /* Write PVD collection index file */
+    sprintf(pvd_filename, "%s.pvd", setname);
+    pvd_fp = fopen(pvd_filename, "w");
+    if (pvd_fp)
+    {
+      fprintf(pvd_fp, "<?xml version=\"1.0\"?>\n");
+      fprintf(pvd_fp, "<VTKFile type=\"Collection\" version=\"0.1\" byte_order=\"LittleEndian\">\n");
+      fprintf(pvd_fp, "  <Collection>\n");
+    }
+
+    /* Write individual step files */
+    for (i = 0; i < num_steps; i++)
+    {
+      sprintf(vtu_filename, "%s_step_%03d.vtu", setname, step_list[i]);
+      write_single_vtu_file(vtu_filename, step_list[i], anz, node, elem, set, setNr,
+                            lcase, num_pts, pts_nodenr, nodeMap, num_elems, elem_indices);
+
+      if (pvd_fp)
+      {
+        fprintf(pvd_fp, "    <DataSet timestep=\"%.6f\" group=\"\" part=\"0\" file=\"%s\"/>\n",
+                step_times[i], vtu_filename);
+      }
+    }
+
+    if (pvd_fp)
+    {
+      fprintf(pvd_fp, "  </Collection>\n");
+      fprintf(pvd_fp, "</VTKFile>\n");
+      fclose(pvd_fp);
+      printf("  -> Written PVD collection: %s (%d steps)\n", pvd_filename, num_steps);
+    }
+
+    free(step_list);
+    free(step_times);
+  }
+  else
+  {
+    /* Single active step / mesh */
+    sprintf(vtu_filename, "%s.vtu", setname);
+    write_single_vtu_file(vtu_filename, 0, anz, node, elem, set, setNr,
+                          lcase, num_pts, pts_nodenr, nodeMap, num_elems, elem_indices);
+  }
+
+  printf("=== VTK/VTU Export Complete ===\n\n");
+
+  free(nodeMap);
+  free(pts_nodenr);
+  return 1;
+}
-- 
2.39.3 (Apple Git-146)


From 1d9dc87476eabc2d191c12db0ab063c07f7a6835 Mon Sep 17 00:00:00 2001
From: Carlo Monjaraz <carlomonjaraztec@proton.me>
Date: Thu, 20 Aug 2026 23:31:50 +0200
Subject: [PATCH 2/4] Fix VTK quadratic element node ordering permutations for
 Hex20 and Penta15

---
 cgx_2.23/src/write2vtu.c | 21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/cgx_2.23/src/write2vtu.c b/cgx_2.23/src/write2vtu.c
index 73055b0..96be005 100644
--- a/cgx_2.23/src/write2vtu.c
+++ b/cgx_2.23/src/write2vtu.c
@@ -85,6 +85,21 @@ static void calc_principal_and_mises(double sxx, double syy, double szz,
   *p3 = r3;
 }
 
+/* Node permutation map from CGX element node ordering to VTK standard ordering */
+static const int vtk_map_hex20[20] = {
+  0, 1, 2, 3, 4, 5, 6, 7,        /* Corners: 0-7 */
+  8, 9, 10, 11,                  /* Bottom edges: 0-1, 1-2, 2-3, 3-0 */
+  16, 17, 18, 19,                /* Top edges in VTK (CGX 16..19) */
+  12, 13, 14, 15                 /* Vertical edges in VTK (CGX 12..15) */
+};
+
+static const int vtk_map_penta15[15] = {
+  0, 1, 2, 3, 4, 5,              /* Corners: bottom 0-2, top 3-5 */
+  6, 7, 8,                       /* Bottom edges */
+  12, 13, 14,                    /* Top edges in VTK (CGX 12..14) */
+  9, 10, 11                      /* Vertical edges in VTK (CGX 9..11) */
+};
+
 /* Map CGX element types to VTK cell types */
 static int get_vtk_cell_type(int cgx_type, int *num_nodes)
 {
@@ -154,7 +169,11 @@ static int write_single_vtu_file(const char *filename, int target_step,
     fprintf(fp, "         ");
     for (j = 0; j < n_nodes; j++)
     {
-      nid = elem[eid].nod[j];
+      int map_idx = j;
+      if (elem[eid].type == 4) map_idx = vtk_map_hex20[j];
+      else if (elem[eid].type == 5) map_idx = vtk_map_penta15[j];
+
+      nid = elem[eid].nod[map_idx];
       fprintf(fp, " %d", (nid >= 0 && nid <= anz->nmax) ? nodeMap[nid] : 0);
     }
     fprintf(fp, "\n");
-- 
2.39.3 (Apple Git-146)


From 52fe456ca09561ebcd46c9e27fd8d60a23279bf8 Mon Sep 17 00:00:00 2001
From: Carlo Monjaraz <carlomonjaraztec@proton.me>
Date: Fri, 21 Aug 2026 00:00:01 +0200
Subject: [PATCH 3/4] Fix VTK quadratic element permutations, physical step
 grouping, and FieldData metadata

- Add node permutations for Hex20 (C3D20) and Penta15 (C3D15) quadratic elements
- Group multi-record dataset fields into unified single-step files
- Add global FieldData under UnstructuredGrid for Frequency_Hz and Mode_Number
- Add Frequency_Hz to PointData array for direct toolbar/coloring access in ParaView
- Support both .vtk and .vtu extension based on requested format
---
 cgx_2.23/src/cgx.h          |   2 +-
 cgx_2.23/src/setFunktions.c |   4 +-
 cgx_2.23/src/write2vtu.c    | 224 +++++++++++++++++++++++++++---------
 3 files changed, 171 insertions(+), 59 deletions(-)

diff --git a/cgx_2.23/src/cgx.h b/cgx_2.23/src/cgx.h
index c5b3aeb..530669e 100755
--- a/cgx_2.23/src/cgx.h
+++ b/cgx_2.23/src/cgx.h
@@ -407,7 +407,7 @@ void splitBiasDiv(int *ptrdiv, double *ptrbias);
 int improveBadTr3(char *setname);
 void printHash();
 int write2foam(char *setname, int strings, char **string, Summen *anz, Nodes *node, Faces *face, Elements *e_enqire, Sets *set, Datasets *lcase );
-int write2vtu(char *setname, int strings, char **string, Summen *anz, Nodes *node, Faces *face, Elements *e_enqire, Sets *set, Datasets *lcase );
+int write2vtu(char *setname, char *format, int strings, char **string, Summen *anz, Nodes *node, Faces *face, Elements *e_enqire, Sets *set, Datasets *lcase );
 int write2dolfyn(char *setname, int strings, char **string, Summen *anz, Nodes *node, Faces *face, Elements *e_enqire, Sets *set);
 int readFoam(char *datin, Summen *anz, Sets **sptr, Nodes **nptr, Elements **eptr, Datasets **lptr );
 int readVtk(char *datin, Summen *anz, Nodes **nptr, Elements **eptr, Datasets **lptr );
diff --git a/cgx_2.23/src/setFunktions.c b/cgx_2.23/src/setFunktions.c
index ec2f89f..7922958 100644
--- a/cgx_2.23/src/setFunktions.c
+++ b/cgx_2.23/src/setFunktions.c
@@ -23733,8 +23733,8 @@ void pre_write( char *record )
   {
     length= strsplt( record, ' ', &dat);
     descalNodes ( anz->n, node, scale);
-    if(length>2) write2vtu( setname, length-2, &dat[2], anz, node, face, e_enqire, set, lcase );
-    else write2vtu( setname, 0, NULL, anz, node, face, e_enqire, set, lcase );
+    if(length>2) write2vtu( setname, format, length-2, &dat[2], anz, node, face, e_enqire, set, lcase );
+    else write2vtu( setname, format, 0, NULL, anz, node, face, e_enqire, set, lcase );
     scalNodes ( anz->n, node, scale);
 
     /* free dat */
diff --git a/cgx_2.23/src/write2vtu.c b/cgx_2.23/src/write2vtu.c
index 96be005..1b1b200 100644
--- a/cgx_2.23/src/write2vtu.c
+++ b/cgx_2.23/src/write2vtu.c
@@ -28,6 +28,21 @@ extern int cur_lc;
 int readfrdblock( int lc, Summen *anz, Nodes *node, Datasets *lcase );
 void calcDatasets( int num_olc, Summen *anz, Nodes *node, Datasets *lcase );
 
+/* Node permutation map from CGX element node ordering to VTK standard ordering */
+static const int vtk_map_hex20[20] = {
+  0, 1, 2, 3, 4, 5, 6, 7,        /* Corners: 0-7 */
+  8, 9, 10, 11,                  /* Bottom edges: 0-1, 1-2, 2-3, 3-0 */
+  16, 17, 18, 19,                /* Top edges in VTK (CGX 16..19) */
+  12, 13, 14, 15                 /* Vertical edges in VTK (CGX 12..15) */
+};
+
+static const int vtk_map_penta15[15] = {
+  0, 1, 2, 3, 4, 5,              /* Corners: bottom 0-2, top 3-5 */
+  6, 7, 8,                       /* Bottom edges */
+  12, 13, 14,                    /* Top edges in VTK (CGX 12..14) */
+  9, 10, 11                      /* Vertical edges in VTK (CGX 9..11) */
+};
+
 /* Calculate Principal Stresses P1 >= P2 >= P3 and von Mises stress analytically */
 static void calc_principal_and_mises(double sxx, double syy, double szz,
                                      double sxy, double syz, double szx,
@@ -85,21 +100,6 @@ static void calc_principal_and_mises(double sxx, double syy, double szz,
   *p3 = r3;
 }
 
-/* Node permutation map from CGX element node ordering to VTK standard ordering */
-static const int vtk_map_hex20[20] = {
-  0, 1, 2, 3, 4, 5, 6, 7,        /* Corners: 0-7 */
-  8, 9, 10, 11,                  /* Bottom edges: 0-1, 1-2, 2-3, 3-0 */
-  16, 17, 18, 19,                /* Top edges in VTK (CGX 16..19) */
-  12, 13, 14, 15                 /* Vertical edges in VTK (CGX 12..15) */
-};
-
-static const int vtk_map_penta15[15] = {
-  0, 1, 2, 3, 4, 5,              /* Corners: bottom 0-2, top 3-5 */
-  6, 7, 8,                       /* Bottom edges */
-  12, 13, 14,                    /* Top edges in VTK (CGX 12..14) */
-  9, 10, 11                      /* Vertical edges in VTK (CGX 9..11) */
-};
-
 /* Map CGX element types to VTK cell types */
 static int get_vtk_cell_type(int cgx_type, int *num_nodes)
 {
@@ -121,8 +121,10 @@ static int get_vtk_cell_type(int cgx_type, int *num_nodes)
   }
 }
 
-/* Write a single .vtu file for a specific step or current dataset */
-static int write_single_vtu_file(const char *filename, int target_step,
+/* Write a single .vtu file for a given physical step */
+static int write_single_vtu_file(const char *filename,
+                                 int num_step_lcs, int *step_lc_indices,
+                                 double step_val, int step_num, int analysis_type,
                                  Summen *anz, Nodes *node, Elements *elem,
                                  Sets *set, int setNr, Datasets *lcase,
                                  int num_pts, int *pts_nodenr, int *nodeMap,
@@ -144,6 +146,34 @@ static int write_single_vtu_file(const char *filename, int target_step,
   fprintf(fp, "<?xml version=\"1.0\"?>\n");
   fprintf(fp, "<VTKFile type=\"UnstructuredGrid\" version=\"0.1\" byte_order=\"LittleEndian\">\n");
   fprintf(fp, "  <UnstructuredGrid>\n");
+
+  /* Global / Step Metadata (FieldData placed directly under UnstructuredGrid) */
+  fprintf(fp, "    <FieldData>\n");
+  if (analysis_type >= 2)
+  {
+    /* Modal / Frequency Analysis */
+    fprintf(fp, "      <DataArray type=\"Float64\" Name=\"Frequency_Hz\" NumberOfTuples=\"1\" format=\"ascii\">\n");
+    fprintf(fp, "        %.6f\n", step_val);
+    fprintf(fp, "      </DataArray>\n");
+    fprintf(fp, "      <DataArray type=\"Int32\" Name=\"Mode_Number\" NumberOfTuples=\"1\" format=\"ascii\">\n");
+    fprintf(fp, "        %d\n", step_num);
+    fprintf(fp, "      </DataArray>\n");
+    fprintf(fp, "      <DataArray type=\"Float64\" Name=\"Angular_Frequency_rad_s\" NumberOfTuples=\"1\" format=\"ascii\">\n");
+    fprintf(fp, "        %.6f\n", 2.0 * M_PI * step_val);
+    fprintf(fp, "      </DataArray>\n");
+  }
+  else
+  {
+    /* Static / Transient Analysis */
+    fprintf(fp, "      <DataArray type=\"Float64\" Name=\"Time\" NumberOfTuples=\"1\" format=\"ascii\">\n");
+    fprintf(fp, "        %.6f\n", step_val);
+    fprintf(fp, "      </DataArray>\n");
+    fprintf(fp, "      <DataArray type=\"Int32\" Name=\"Step_Number\" NumberOfTuples=\"1\" format=\"ascii\">\n");
+    fprintf(fp, "        %d\n", step_num);
+    fprintf(fp, "      </DataArray>\n");
+  }
+  fprintf(fp, "    </FieldData>\n");
+
   fprintf(fp, "    <Piece NumberOfPoints=\"%d\" NumberOfCells=\"%d\">\n", num_pts, num_elems);
 
   /* 1. Point Coordinates */
@@ -205,16 +235,12 @@ static int write_single_vtu_file(const char *filename, int target_step,
 
   /* 3. Point Data Fields */
   fprintf(fp, "      <PointData>\n");
-  if (anz->l > 0 && lcase != NULL)
+  if (anz->l > 0 && lcase != NULL && num_step_lcs > 0)
   {
-    for (lc = 0; lc < anz->l; lc++)
+    for (k = 0; k < num_step_lcs; k++)
     {
-      /* Filter datasets by target_step if specified (target_step > 0) */
-      if (target_step > 0 && lcase[lc].step_number != target_step) continue;
-      if (target_step == 0 && lc != cur_lc && anz->l > 1)
-      {
-        if (lcase[lc].step_number != lcase[cur_lc].step_number) continue;
-      }
+      lc = step_lc_indices[k];
+      if (lc < 0 || lc >= anz->l) continue;
 
       /* Ensure dataset is loaded in memory */
       if (!lcase[lc].loaded)
@@ -309,6 +335,17 @@ static int write_single_vtu_file(const char *filename, int target_step,
         fprintf(fp, "        </DataArray>\n");
       }
     }
+
+    /* Include Frequency_Hz as accessible PointData field for modal analysis */
+    if (analysis_type >= 2)
+    {
+      fprintf(fp, "        <DataArray type=\"Float32\" Name=\"Frequency_Hz\" NumberOfComponents=\"1\" format=\"ascii\">\n");
+      for (i = 0; i < num_pts; i++)
+      {
+        fprintf(fp, "          %.6e\n", step_val);
+      }
+      fprintf(fp, "        </DataArray>\n");
+    }
   }
   fprintf(fp, "      </PointData>\n");
 
@@ -340,8 +377,18 @@ static int write_single_vtu_file(const char *filename, int target_step,
   return 1;
 }
 
-/* Main entry point for send <set> vtu [all] */
-int write2vtu(char *setname, int strings, char **string, Summen *anz,
+/* Structure to group datasets belonging to the same physical step/frame */
+typedef struct {
+  int    step_number;
+  int    analysis_type;
+  double value;
+  char   name[MAX_LINE_LENGTH];
+  int    num_lcs;
+  int    lcs[64];
+} PhysStep;
+
+/* Main entry point for send <set> vtu/vtk [all] */
+int write2vtu(char *setname, char *format, int strings, char **string, Summen *anz,
               Nodes *node, Faces *face, Elements *elem, Sets *set, Datasets *lcase)
 {
   int setNr, i, j, k, eid, nid, n_nodes, vtk_type;
@@ -350,12 +397,18 @@ int write2vtu(char *setname, int strings, char **string, Summen *anz,
   int *pts_nodenr = NULL;
   int *elem_indices = NULL;
   int export_all_steps = 0;
-  int num_steps = 0, *step_list = NULL;
-  double *step_times = NULL;
+  char ext[8] = "vtu";
   char vtu_filename[MAX_LINE_LENGTH];
   char pvd_filename[MAX_LINE_LENGTH];
   FILE *pvd_fp = NULL;
 
+  PhysStep *psteps = NULL;
+  int num_psteps = 0;
+  int active_pstep = 0;
+
+  if (format != NULL && compare(format, "vtk", 3) == 3) strcpy(ext, "vtk");
+  else if (format != NULL && compare(format, "vtu", 3) == 3) strcpy(ext, "vtu");
+
   if (anz->n == 0 || anz->e == 0)
   {
     printf("ERROR: No mesh available to export.\n");
@@ -378,7 +431,7 @@ int write2vtu(char *setname, int strings, char **string, Summen *anz,
     }
   }
 
-  printf("\n=== Exporting to VTK XML Unstructured Grid (.vtu) ===\n");
+  printf("\n=== Exporting to VTK XML Unstructured Grid (.%s) ===\n", ext);
   printf("Set: %s (mode: %s)\n", setname, export_all_steps ? "all steps (.pvd)" : "active step (.vtu)");
 
   /* 1. Identify Elements in Set */
@@ -427,28 +480,60 @@ int write2vtu(char *setname, int strings, char **string, Summen *anz,
     }
   }
 
-  /* 3. Export Data */
-  if (export_all_steps && anz->l > 0)
+  /* 3. Group Datasets into Physical Steps / Frames */
+  if (anz->l > 0 && lcase != NULL)
   {
-    /* Find unique steps */
-    step_list = (int *)malloc(anz->l * sizeof(int));
-    step_times = (double *)malloc(anz->l * sizeof(double));
+    psteps = (PhysStep *)calloc(anz->l, sizeof(PhysStep));
     for (i = 0; i < anz->l; i++)
     {
-      int step = lcase[i].step_number;
-      int exists = 0;
-      for (j = 0; j < num_steps; j++)
+      int matched = -1;
+
+      for (j = 0; j < num_psteps; j++)
+      {
+        /* Datasets with same step_number belong to the same physical step/mode */
+        if (psteps[j].step_number == lcase[i].step_number)
+        {
+          matched = j;
+          break;
+        }
+      }
+
+      if (matched >= 0)
+      {
+        if (psteps[matched].num_lcs < 64)
+        {
+          psteps[matched].lcs[psteps[matched].num_lcs++] = i;
+        }
+      }
+      else
       {
-        if (step_list[j] == step) { exists = 1; break; }
+        psteps[num_psteps].step_number = lcase[i].step_number;
+        psteps[num_psteps].analysis_type = lcase[i].analysis_type;
+        psteps[num_psteps].value = (double)lcase[i].value;
+        strcpy(psteps[num_psteps].name, lcase[i].analysis_name);
+        psteps[num_psteps].lcs[0] = i;
+        psteps[num_psteps].num_lcs = 1;
+        num_psteps++;
       }
-      if (!exists)
+    }
+
+    /* Find which physical step contains the current active dataset */
+    for (j = 0; j < num_psteps; j++)
+    {
+      for (k = 0; k < psteps[j].num_lcs; k++)
       {
-        step_list[num_steps] = step;
-        step_times[num_steps] = (double)lcase[i].value;
-        num_steps++;
+        if (psteps[j].lcs[k] == cur_lc)
+        {
+          active_pstep = j;
+          break;
+        }
       }
     }
+  }
 
+  /* 4. Export Data */
+  if (export_all_steps && num_psteps > 0)
+  {
     /* Write PVD collection index file */
     sprintf(pvd_filename, "%s.pvd", setname);
     pvd_fp = fopen(pvd_filename, "w");
@@ -459,17 +544,34 @@ int write2vtu(char *setname, int strings, char **string, Summen *anz,
       fprintf(pvd_fp, "  <Collection>\n");
     }
 
-    /* Write individual step files */
-    for (i = 0; i < num_steps; i++)
+    /* Write individual physical step files */
+    for (i = 0; i < num_psteps; i++)
     {
-      sprintf(vtu_filename, "%s_step_%03d.vtu", setname, step_list[i]);
-      write_single_vtu_file(vtu_filename, step_list[i], anz, node, elem, set, setNr,
-                            lcase, num_pts, pts_nodenr, nodeMap, num_elems, elem_indices);
+      sprintf(vtu_filename, "%s_step_%03d.%s", setname, i + 1, ext);
+      write_single_vtu_file(vtu_filename, psteps[i].num_lcs, psteps[i].lcs,
+                            psteps[i].value, psteps[i].step_number, psteps[i].analysis_type,
+                            anz, node, elem, set, setNr, lcase,
+                            num_pts, pts_nodenr, nodeMap, num_elems, elem_indices);
 
       if (pvd_fp)
       {
+        /* In ParaView PVD, each frame MUST have a unique timestep value.
+           If time values are not strictly increasing (e.g. modal analysis with duplicate frequencies),
+           use the 1-based step index to ensure ParaView loads each mode as an independent frame. */
+        double tval = (double)(i + 1);
+        if (num_psteps > 1 && psteps[num_psteps - 1].value > psteps[0].value)
+        {
+          /* Check if all time values are strictly increasing */
+          int strictly_increasing = 1;
+          for (j = 1; j < num_psteps; j++)
+          {
+            if (psteps[j].value <= psteps[j - 1].value) { strictly_increasing = 0; break; }
+          }
+          if (strictly_increasing) tval = psteps[i].value;
+        }
+
         fprintf(pvd_fp, "    <DataSet timestep=\"%.6f\" group=\"\" part=\"0\" file=\"%s\"/>\n",
-                step_times[i], vtu_filename);
+                tval, vtu_filename);
       }
     }
 
@@ -478,22 +580,32 @@ int write2vtu(char *setname, int strings, char **string, Summen *anz,
       fprintf(pvd_fp, "  </Collection>\n");
       fprintf(pvd_fp, "</VTKFile>\n");
       fclose(pvd_fp);
-      printf("  -> Written PVD collection: %s (%d steps)\n", pvd_filename, num_steps);
+      printf("  -> Written PVD collection: %s (%d physical steps)\n", pvd_filename, num_psteps);
     }
-
-    free(step_list);
-    free(step_times);
   }
   else
   {
     /* Single active step / mesh */
-    sprintf(vtu_filename, "%s.vtu", setname);
-    write_single_vtu_file(vtu_filename, 0, anz, node, elem, set, setNr,
-                          lcase, num_pts, pts_nodenr, nodeMap, num_elems, elem_indices);
+    sprintf(vtu_filename, "%s.%s", setname, ext);
+    if (num_psteps > 0)
+    {
+      write_single_vtu_file(vtu_filename, psteps[active_pstep].num_lcs, psteps[active_pstep].lcs,
+                            psteps[active_pstep].value, psteps[active_pstep].step_number, psteps[active_pstep].analysis_type,
+                            anz, node, elem, set, setNr, lcase,
+                            num_pts, pts_nodenr, nodeMap, num_elems, elem_indices);
+    }
+    else
+    {
+      write_single_vtu_file(vtu_filename, 0, NULL,
+                            0.0, 0, 0,
+                            anz, node, elem, set, setNr, lcase,
+                            num_pts, pts_nodenr, nodeMap, num_elems, elem_indices);
+    }
   }
 
   printf("=== VTK/VTU Export Complete ===\n\n");
 
+  if (psteps) free(psteps);
   free(nodeMap);
   free(pts_nodenr);
   return 1;
-- 
2.39.3 (Apple Git-146)


From 3b8e2fe726bf5c2b8ea198a8702ac629a41aff4d Mon Sep 17 00:00:00 2001
From: Carlo Monjaraz <carlomonjaraztec@proton.me>
Date: Fri, 21 Aug 2026 00:16:19 +0200
Subject: [PATCH 4/4] Implement native Base64 binary VTU export with optional
 ASCII flag

- Default to compact, fast Base64 inline binary encoding for VTU files
- Support optional 'ascii' flag (send <set> vtu [all] [ascii])
- Enforce strict VTK XML canonical tag ordering (<PointData>, <CellData>, <Points>, <Cells>)
- Add explicit header_type='UInt32' and Int32 cell connectivity/offsets for ParaView compatibility
- Remove redundant vtk format alias in favor of standardized vtu
---
 cgx_2.23/src/cgx.h          |   2 +-
 cgx_2.23/src/setFunktions.c |   6 +-
 cgx_2.23/src/write2vtu.c    | 642 +++++++++++++++++++++++++++---------
 3 files changed, 489 insertions(+), 161 deletions(-)

diff --git a/cgx_2.23/src/cgx.h b/cgx_2.23/src/cgx.h
index 530669e..c5b3aeb 100755
--- a/cgx_2.23/src/cgx.h
+++ b/cgx_2.23/src/cgx.h
@@ -407,7 +407,7 @@ void splitBiasDiv(int *ptrdiv, double *ptrbias);
 int improveBadTr3(char *setname);
 void printHash();
 int write2foam(char *setname, int strings, char **string, Summen *anz, Nodes *node, Faces *face, Elements *e_enqire, Sets *set, Datasets *lcase );
-int write2vtu(char *setname, char *format, int strings, char **string, Summen *anz, Nodes *node, Faces *face, Elements *e_enqire, Sets *set, Datasets *lcase );
+int write2vtu(char *setname, int strings, char **string, Summen *anz, Nodes *node, Faces *face, Elements *e_enqire, Sets *set, Datasets *lcase );
 int write2dolfyn(char *setname, int strings, char **string, Summen *anz, Nodes *node, Faces *face, Elements *e_enqire, Sets *set);
 int readFoam(char *datin, Summen *anz, Sets **sptr, Nodes **nptr, Elements **eptr, Datasets **lptr );
 int readVtk(char *datin, Summen *anz, Nodes **nptr, Elements **eptr, Datasets **lptr );
diff --git a/cgx_2.23/src/setFunktions.c b/cgx_2.23/src/setFunktions.c
index 7922958..34eb948 100644
--- a/cgx_2.23/src/setFunktions.c
+++ b/cgx_2.23/src/setFunktions.c
@@ -23729,12 +23729,12 @@ void pre_write( char *record )
     /* free dat */
     for(i=0; i<length; i++) free(dat[i]); free(dat);
   }
-  else if (compare( format, "vtu", 3)== 3 || compare( format, "vtk", 3)== 3)
+  else if (compare( format, "vtu", 3)== 3)
   {
     length= strsplt( record, ' ', &dat);
     descalNodes ( anz->n, node, scale);
-    if(length>2) write2vtu( setname, format, length-2, &dat[2], anz, node, face, e_enqire, set, lcase );
-    else write2vtu( setname, format, 0, NULL, anz, node, face, e_enqire, set, lcase );
+    if(length>2) write2vtu( setname, length-2, &dat[2], anz, node, face, e_enqire, set, lcase );
+    else write2vtu( setname, 0, NULL, anz, node, face, e_enqire, set, lcase );
     scalNodes ( anz->n, node, scale);
 
     /* free dat */
diff --git a/cgx_2.23/src/write2vtu.c b/cgx_2.23/src/write2vtu.c
index 1b1b200..02f473e 100644
--- a/cgx_2.23/src/write2vtu.c
+++ b/cgx_2.23/src/write2vtu.c
@@ -12,12 +12,13 @@
 /* --------------------------------------------------------------------  */
 
 /* write2vtu.c - Export mesh and field datasets to VTK XML Unstructured Grid (.vtu) */
-/* and ParaView Data collection (.pvd)                                               */
+/* and ParaView Data collection (.pvd) with Base64 Binary (default) & ASCII support  */
 
 #include <cgx.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <stdint.h>
 #include <math.h>
 
 #ifndef M_PI
@@ -28,6 +29,47 @@ extern int cur_lc;
 int readfrdblock( int lc, Summen *anz, Nodes *node, Datasets *lcase );
 void calcDatasets( int num_olc, Summen *anz, Nodes *node, Datasets *lcase );
 
+/* Base64 encoding table */
+static const char b64_table[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+
+/* Write a binary buffer as a Base64-encoded VTK data block (32-bit length header + payload) */
+static void write_base64_data_array(FILE *fp, const void *data, uint32_t num_bytes)
+{
+  uint32_t total_len = sizeof(uint32_t) + num_bytes;
+  unsigned char *buf = (unsigned char *)malloc(total_len);
+  uint32_t i;
+
+  if (!buf)
+  {
+    printf("ERROR: Out of memory in write_base64_data_array\n");
+    return;
+  }
+
+  /* Copy 32-bit length header in native little-endian */
+  memcpy(buf, &num_bytes, sizeof(uint32_t));
+  /* Copy payload */
+  if (num_bytes > 0 && data != NULL)
+  {
+    memcpy(buf + sizeof(uint32_t), data, num_bytes);
+  }
+
+  fprintf(fp, "\n          ");
+  for (i = 0; i < total_len; i += 3)
+  {
+    uint32_t b0 = buf[i];
+    uint32_t b1 = (i + 1 < total_len) ? buf[i + 1] : 0;
+    uint32_t b2 = (i + 2 < total_len) ? buf[i + 2] : 0;
+    uint32_t triple = (b0 << 16) | (b1 << 8) | b2;
+
+    fputc(b64_table[(triple >> 18) & 0x3F], fp);
+    fputc(b64_table[(triple >> 12) & 0x3F], fp);
+    fputc((i + 1 < total_len) ? b64_table[(triple >> 6) & 0x3F] : '=', fp);
+    fputc((i + 2 < total_len) ? b64_table[triple & 0x3F] : '=', fp);
+  }
+  fprintf(fp, "\n        ");
+  free(buf);
+}
+
 /* Node permutation map from CGX element node ordering to VTK standard ordering */
 static const int vtk_map_hex20[20] = {
   0, 1, 2, 3, 4, 5, 6, 7,        /* Corners: 0-7 */
@@ -125,6 +167,7 @@ static int get_vtk_cell_type(int cgx_type, int *num_nodes)
 static int write_single_vtu_file(const char *filename,
                                  int num_step_lcs, int *step_lc_indices,
                                  double step_val, int step_num, int analysis_type,
+                                 int is_binary,
                                  Summen *anz, Nodes *node, Elements *elem,
                                  Sets *set, int setNr, Datasets *lcase,
                                  int num_pts, int *pts_nodenr, int *nodeMap,
@@ -136,6 +179,14 @@ static int write_single_vtu_file(const char *filename,
   double sxx, syy, szz, sxy, syz, szx;
   double p1, p2, p3, vm;
 
+  /* Buffers for binary encoding */
+  double *f64_buf = NULL;
+  float  *f32_buf = NULL;
+  int64_t *i64_buf = NULL;
+  int32_t *i32_buf = NULL;
+  uint8_t *u8_buf = NULL;
+  int total_conn = 0;
+
   fp = fopen(filename, "w");
   if (!fp)
   {
@@ -144,7 +195,7 @@ static int write_single_vtu_file(const char *filename,
   }
 
   fprintf(fp, "<?xml version=\"1.0\"?>\n");
-  fprintf(fp, "<VTKFile type=\"UnstructuredGrid\" version=\"0.1\" byte_order=\"LittleEndian\">\n");
+  fprintf(fp, "<VTKFile type=\"UnstructuredGrid\" version=\"0.1\" byte_order=\"LittleEndian\" header_type=\"UInt32\">\n");
   fprintf(fp, "  <UnstructuredGrid>\n");
 
   /* Global / Step Metadata (FieldData placed directly under UnstructuredGrid) */
@@ -152,89 +203,48 @@ static int write_single_vtu_file(const char *filename,
   if (analysis_type >= 2)
   {
     /* Modal / Frequency Analysis */
-    fprintf(fp, "      <DataArray type=\"Float64\" Name=\"Frequency_Hz\" NumberOfTuples=\"1\" format=\"ascii\">\n");
-    fprintf(fp, "        %.6f\n", step_val);
-    fprintf(fp, "      </DataArray>\n");
-    fprintf(fp, "      <DataArray type=\"Int32\" Name=\"Mode_Number\" NumberOfTuples=\"1\" format=\"ascii\">\n");
-    fprintf(fp, "        %d\n", step_num);
-    fprintf(fp, "      </DataArray>\n");
-    fprintf(fp, "      <DataArray type=\"Float64\" Name=\"Angular_Frequency_rad_s\" NumberOfTuples=\"1\" format=\"ascii\">\n");
-    fprintf(fp, "        %.6f\n", 2.0 * M_PI * step_val);
-    fprintf(fp, "      </DataArray>\n");
+    double ang_freq = 2.0 * M_PI * step_val;
+    int32_t m_num = step_num;
+    fprintf(fp, "      <DataArray type=\"Float64\" Name=\"Frequency_Hz\" NumberOfTuples=\"1\" format=\"%s\">",
+            is_binary ? "binary" : "ascii");
+    if (is_binary) write_base64_data_array(fp, &step_val, sizeof(double));
+    else fprintf(fp, "\n          %.6f\n        ", step_val);
+    fprintf(fp, "</DataArray>\n");
+
+    fprintf(fp, "      <DataArray type=\"Int32\" Name=\"Mode_Number\" NumberOfTuples=\"1\" format=\"%s\">",
+            is_binary ? "binary" : "ascii");
+    if (is_binary) write_base64_data_array(fp, &m_num, sizeof(int32_t));
+    else fprintf(fp, "\n          %d\n        ", m_num);
+    fprintf(fp, "</DataArray>\n");
+
+    fprintf(fp, "      <DataArray type=\"Float64\" Name=\"Angular_Frequency_rad_s\" NumberOfTuples=\"1\" format=\"%s\">",
+            is_binary ? "binary" : "ascii");
+    if (is_binary) write_base64_data_array(fp, &ang_freq, sizeof(double));
+    else fprintf(fp, "\n          %.6f\n        ", ang_freq);
+    fprintf(fp, "</DataArray>\n");
   }
   else
   {
     /* Static / Transient Analysis */
-    fprintf(fp, "      <DataArray type=\"Float64\" Name=\"Time\" NumberOfTuples=\"1\" format=\"ascii\">\n");
-    fprintf(fp, "        %.6f\n", step_val);
-    fprintf(fp, "      </DataArray>\n");
-    fprintf(fp, "      <DataArray type=\"Int32\" Name=\"Step_Number\" NumberOfTuples=\"1\" format=\"ascii\">\n");
-    fprintf(fp, "        %d\n", step_num);
-    fprintf(fp, "      </DataArray>\n");
+    int32_t s_num = step_num;
+    fprintf(fp, "      <DataArray type=\"Float64\" Name=\"Time\" NumberOfTuples=\"1\" format=\"%s\">",
+            is_binary ? "binary" : "ascii");
+    if (is_binary) write_base64_data_array(fp, &step_val, sizeof(double));
+    else fprintf(fp, "\n          %.6f\n        ", step_val);
+    fprintf(fp, "</DataArray>\n");
+
+    fprintf(fp, "      <DataArray type=\"Int32\" Name=\"Step_Number\" NumberOfTuples=\"1\" format=\"%s\">",
+            is_binary ? "binary" : "ascii");
+    if (is_binary) write_base64_data_array(fp, &s_num, sizeof(int32_t));
+    else fprintf(fp, "\n          %d\n        ", s_num);
+    fprintf(fp, "</DataArray>\n");
   }
   fprintf(fp, "    </FieldData>\n");
 
   fprintf(fp, "    <Piece NumberOfPoints=\"%d\" NumberOfCells=\"%d\">\n", num_pts, num_elems);
-
-  /* 1. Point Coordinates */
-  fprintf(fp, "      <Points>\n");
-  fprintf(fp, "        <DataArray type=\"Float64\" Name=\"Points\" NumberOfComponents=\"3\" format=\"ascii\">\n");
-  for (i = 0; i < num_pts; i++)
-  {
-    nid = pts_nodenr[i];
-    fprintf(fp, "          %.9e %.9e %.9e\n", node[nid].nx, node[nid].ny, node[nid].nz);
-  }
-  fprintf(fp, "        </DataArray>\n");
-  fprintf(fp, "      </Points>\n");
-
-  /* 2. Cells (Topology, Offsets, Types) */
-  fprintf(fp, "      <Cells>\n");
-
-  /* Connectivity */
-  fprintf(fp, "        <DataArray type=\"Int64\" Name=\"connectivity\" format=\"ascii\">\n");
-  for (i = 0; i < num_elems; i++)
-  {
-    eid = elem_indices[i];
-    vtk_type = get_vtk_cell_type(elem[eid].type, &n_nodes);
-    fprintf(fp, "         ");
-    for (j = 0; j < n_nodes; j++)
-    {
-      int map_idx = j;
-      if (elem[eid].type == 4) map_idx = vtk_map_hex20[j];
-      else if (elem[eid].type == 5) map_idx = vtk_map_penta15[j];
-
-      nid = elem[eid].nod[map_idx];
-      fprintf(fp, " %d", (nid >= 0 && nid <= anz->nmax) ? nodeMap[nid] : 0);
-    }
-    fprintf(fp, "\n");
-  }
-  fprintf(fp, "        </DataArray>\n");
-
-  /* Offsets */
-  fprintf(fp, "        <DataArray type=\"Int64\" Name=\"offsets\" format=\"ascii\">\n");
-  offset = 0;
-  for (i = 0; i < num_elems; i++)
-  {
-    eid = elem_indices[i];
-    get_vtk_cell_type(elem[eid].type, &n_nodes);
-    offset += n_nodes;
-    fprintf(fp, "          %ld\n", offset);
-  }
-  fprintf(fp, "        </DataArray>\n");
-
-  /* Types */
-  fprintf(fp, "        <DataArray type=\"UInt8\" Name=\"types\" format=\"ascii\">\n");
-  for (i = 0; i < num_elems; i++)
-  {
-    eid = elem_indices[i];
-    vtk_type = get_vtk_cell_type(elem[eid].type, &n_nodes);
-    fprintf(fp, "          %d\n", vtk_type);
-  }
-  fprintf(fp, "        </DataArray>\n");
-  fprintf(fp, "      </Cells>\n");
-
-  /* 3. Point Data Fields */
+  fprintf(fp, "      <!-- 1. Point Data Fields -->\n");
   fprintf(fp, "      <PointData>\n");
+
   if (anz->l > 0 && lcase != NULL && num_step_lcs > 0)
   {
     for (k = 0; k < num_step_lcs; k++)
@@ -254,126 +264,442 @@ static int write_single_vtu_file(const char *filename,
       /* Displacements Vector (3 components) */
       if (compare(lcase[lc].name, "DISP", 4) == 4 || (lcase[lc].ncomps >= 3 && compare(lcase[lc].compName[0], "D1", 2) == 2))
       {
-        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"Displacements\" NumberOfComponents=\"3\" format=\"ascii\">\n");
-        for (i = 0; i < num_pts; i++)
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"Displacements\" NumberOfComponents=\"3\" format=\"%s\">",
+                is_binary ? "binary" : "ascii");
+        if (is_binary)
+        {
+          f32_buf = (float *)malloc(num_pts * 3 * sizeof(float));
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            f32_buf[i * 3 + 0] = (float)lcase[lc].dat[0][nid];
+            f32_buf[i * 3 + 1] = (float)lcase[lc].dat[1][nid];
+            f32_buf[i * 3 + 2] = (float)lcase[lc].dat[2][nid];
+          }
+          write_base64_data_array(fp, f32_buf, num_pts * 3 * sizeof(float));
+          free(f32_buf);
+        }
+        else
         {
-          nid = pts_nodenr[i];
-          fprintf(fp, "          %.6e %.6e %.6e\n",
-                  lcase[lc].dat[0][nid], lcase[lc].dat[1][nid], lcase[lc].dat[2][nid]);
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            fprintf(fp, "\n          %.6e %.6e %.6e",
+                    lcase[lc].dat[0][nid], lcase[lc].dat[1][nid], lcase[lc].dat[2][nid]);
+          }
+          fprintf(fp, "\n        ");
         }
-        fprintf(fp, "        </DataArray>\n");
+        fprintf(fp, "</DataArray>\n");
       }
       /* Stress Tensor & Precomputed Invariants */
       else if (compare(lcase[lc].name, "STRESS", 6) == 6 || lcase[lc].ncomps >= 6)
       {
         /* 6-component Stress Tensor */
-        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"Stress\" NumberOfComponents=\"6\" format=\"ascii\">\n");
-        for (i = 0; i < num_pts; i++)
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"Stress\" NumberOfComponents=\"6\" format=\"%s\">",
+                is_binary ? "binary" : "ascii");
+        if (is_binary)
         {
-          nid = pts_nodenr[i];
-          fprintf(fp, "          %.6e %.6e %.6e %.6e %.6e %.6e\n",
-                  lcase[lc].dat[0][nid], lcase[lc].dat[1][nid], lcase[lc].dat[2][nid],
-                  lcase[lc].dat[3][nid], lcase[lc].dat[4][nid], lcase[lc].dat[5][nid]);
+          f32_buf = (float *)malloc(num_pts * 6 * sizeof(float));
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            f32_buf[i * 6 + 0] = (float)lcase[lc].dat[0][nid];
+            f32_buf[i * 6 + 1] = (float)lcase[lc].dat[1][nid];
+            f32_buf[i * 6 + 2] = (float)lcase[lc].dat[2][nid];
+            f32_buf[i * 6 + 3] = (float)lcase[lc].dat[3][nid];
+            f32_buf[i * 6 + 4] = (float)lcase[lc].dat[4][nid];
+            f32_buf[i * 6 + 5] = (float)lcase[lc].dat[5][nid];
+          }
+          write_base64_data_array(fp, f32_buf, num_pts * 6 * sizeof(float));
+          free(f32_buf);
         }
-        fprintf(fp, "        </DataArray>\n");
+        else
+        {
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            fprintf(fp, "\n          %.6e %.6e %.6e %.6e %.6e %.6e",
+                    lcase[lc].dat[0][nid], lcase[lc].dat[1][nid], lcase[lc].dat[2][nid],
+                    lcase[lc].dat[3][nid], lcase[lc].dat[4][nid], lcase[lc].dat[5][nid]);
+          }
+          fprintf(fp, "\n        ");
+        }
+        fprintf(fp, "</DataArray>\n");
 
         /* Precomputed von Mises Scalar */
-        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"vonMises\" NumberOfComponents=\"1\" format=\"ascii\">\n");
-        for (i = 0; i < num_pts; i++)
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"vonMises\" NumberOfComponents=\"1\" format=\"%s\">",
+                is_binary ? "binary" : "ascii");
+        if (is_binary)
         {
-          nid = pts_nodenr[i];
-          sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
-          sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
-          calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
-          fprintf(fp, "          %.6e\n", vm);
+          f32_buf = (float *)malloc(num_pts * sizeof(float));
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+            sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+            calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+            f32_buf[i] = (float)vm;
+          }
+          write_base64_data_array(fp, f32_buf, num_pts * sizeof(float));
+          free(f32_buf);
         }
-        fprintf(fp, "        </DataArray>\n");
+        else
+        {
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+            sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+            calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+            fprintf(fp, "\n          %.6e", vm);
+          }
+          fprintf(fp, "\n        ");
+        }
+        fprintf(fp, "</DataArray>\n");
 
         /* Principal Stresses */
-        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"PrincipalStress_1\" NumberOfComponents=\"1\" format=\"ascii\">\n");
-        for (i = 0; i < num_pts; i++)
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"PrincipalStress_1\" NumberOfComponents=\"1\" format=\"%s\">",
+                is_binary ? "binary" : "ascii");
+        if (is_binary)
         {
-          nid = pts_nodenr[i];
-          sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
-          sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
-          calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
-          fprintf(fp, "          %.6e\n", p1);
+          f32_buf = (float *)malloc(num_pts * sizeof(float));
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+            sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+            calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+            f32_buf[i] = (float)p1;
+          }
+          write_base64_data_array(fp, f32_buf, num_pts * sizeof(float));
+          free(f32_buf);
         }
-        fprintf(fp, "        </DataArray>\n");
+        else
+        {
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+            sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+            calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+            fprintf(fp, "\n          %.6e", p1);
+          }
+          fprintf(fp, "\n        ");
+        }
+        fprintf(fp, "</DataArray>\n");
 
-        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"PrincipalStress_2\" NumberOfComponents=\"1\" format=\"ascii\">\n");
-        for (i = 0; i < num_pts; i++)
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"PrincipalStress_2\" NumberOfComponents=\"1\" format=\"%s\">",
+                is_binary ? "binary" : "ascii");
+        if (is_binary)
         {
-          nid = pts_nodenr[i];
-          sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
-          sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
-          calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
-          fprintf(fp, "          %.6e\n", p2);
+          f32_buf = (float *)malloc(num_pts * sizeof(float));
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+            sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+            calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+            f32_buf[i] = (float)p2;
+          }
+          write_base64_data_array(fp, f32_buf, num_pts * sizeof(float));
+          free(f32_buf);
         }
-        fprintf(fp, "        </DataArray>\n");
+        else
+        {
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+            sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+            calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+            fprintf(fp, "\n          %.6e", p2);
+          }
+          fprintf(fp, "\n        ");
+        }
+        fprintf(fp, "</DataArray>\n");
 
-        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"PrincipalStress_3\" NumberOfComponents=\"1\" format=\"ascii\">\n");
-        for (i = 0; i < num_pts; i++)
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"PrincipalStress_3\" NumberOfComponents=\"1\" format=\"%s\">",
+                is_binary ? "binary" : "ascii");
+        if (is_binary)
+        {
+          f32_buf = (float *)malloc(num_pts * sizeof(float));
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+            sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+            calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+            f32_buf[i] = (float)p3;
+          }
+          write_base64_data_array(fp, f32_buf, num_pts * sizeof(float));
+          free(f32_buf);
+        }
+        else
         {
-          nid = pts_nodenr[i];
-          sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
-          sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
-          calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
-          fprintf(fp, "          %.6e\n", p3);
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            sxx = lcase[lc].dat[0][nid]; syy = lcase[lc].dat[1][nid]; szz = lcase[lc].dat[2][nid];
+            sxy = lcase[lc].dat[3][nid]; syz = lcase[lc].dat[4][nid]; szx = lcase[lc].dat[5][nid];
+            calc_principal_and_mises(sxx, syy, szz, sxy, syz, szx, &p1, &p2, &p3, &vm);
+            fprintf(fp, "\n          %.6e", p3);
+          }
+          fprintf(fp, "\n        ");
         }
-        fprintf(fp, "        </DataArray>\n");
+        fprintf(fp, "</DataArray>\n");
       }
       /* Scalar Datasets (Temperature, PEEQ, Pressure, etc.) */
       else if (lcase[lc].ncomps == 1)
       {
-        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"%s\" NumberOfComponents=\"1\" format=\"ascii\">\n", lcase[lc].name);
-        for (i = 0; i < num_pts; i++)
+        fprintf(fp, "        <DataArray type=\"Float32\" Name=\"%s\" NumberOfComponents=\"1\" format=\"%s\">",
+                lcase[lc].name, is_binary ? "binary" : "ascii");
+        if (is_binary)
         {
-          nid = pts_nodenr[i];
-          fprintf(fp, "          %.6e\n", lcase[lc].dat[0][nid]);
+          f32_buf = (float *)malloc(num_pts * sizeof(float));
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            f32_buf[i] = (float)lcase[lc].dat[0][nid];
+          }
+          write_base64_data_array(fp, f32_buf, num_pts * sizeof(float));
+          free(f32_buf);
+        }
+        else
+        {
+          for (i = 0; i < num_pts; i++)
+          {
+            nid = pts_nodenr[i];
+            fprintf(fp, "\n          %.6e", lcase[lc].dat[0][nid]);
+          }
+          fprintf(fp, "\n        ");
         }
-        fprintf(fp, "        </DataArray>\n");
+        fprintf(fp, "</DataArray>\n");
       }
     }
 
     /* Include Frequency_Hz as accessible PointData field for modal analysis */
     if (analysis_type >= 2)
     {
-      fprintf(fp, "        <DataArray type=\"Float32\" Name=\"Frequency_Hz\" NumberOfComponents=\"1\" format=\"ascii\">\n");
-      for (i = 0; i < num_pts; i++)
+      fprintf(fp, "        <DataArray type=\"Float32\" Name=\"Frequency_Hz\" NumberOfComponents=\"1\" format=\"%s\">",
+              is_binary ? "binary" : "ascii");
+      if (is_binary)
       {
-        fprintf(fp, "          %.6e\n", step_val);
+        f32_buf = (float *)malloc(num_pts * sizeof(float));
+        for (i = 0; i < num_pts; i++) f32_buf[i] = (float)step_val;
+        write_base64_data_array(fp, f32_buf, num_pts * sizeof(float));
+        free(f32_buf);
       }
-      fprintf(fp, "        </DataArray>\n");
+      else
+      {
+        for (i = 0; i < num_pts; i++) fprintf(fp, "\n          %.6e", step_val);
+        fprintf(fp, "\n        ");
+      }
+      fprintf(fp, "</DataArray>\n");
     }
   }
   fprintf(fp, "      </PointData>\n");
 
   /* 4. Cell Data (Element IDs & Materials) */
   fprintf(fp, "      <CellData>\n");
-  fprintf(fp, "        <DataArray type=\"Int32\" Name=\"Element_ID\" format=\"ascii\">\n");
-  for (i = 0; i < num_elems; i++)
+  fprintf(fp, "        <DataArray type=\"Int32\" Name=\"Element_ID\" format=\"%s\">",
+          is_binary ? "binary" : "ascii");
+  if (is_binary)
   {
-    eid = elem_indices[i];
-    fprintf(fp, "          %d\n", elem[eid].nr);
+    i32_buf = (int32_t *)malloc(num_elems * sizeof(int32_t));
+    for (i = 0; i < num_elems; i++)
+    {
+      eid = elem_indices[i];
+      i32_buf[i] = elem[eid].nr;
+    }
+    write_base64_data_array(fp, i32_buf, num_elems * sizeof(int32_t));
+    free(i32_buf);
   }
-  fprintf(fp, "        </DataArray>\n");
+  else
+  {
+    for (i = 0; i < num_elems; i++)
+    {
+      eid = elem_indices[i];
+      fprintf(fp, "\n          %d", elem[eid].nr);
+    }
+    fprintf(fp, "\n        ");
+  }
+  fprintf(fp, "</DataArray>\n");
+
+  fprintf(fp, "        <DataArray type=\"Int32\" Name=\"Material_ID\" format=\"%s\">",
+          is_binary ? "binary" : "ascii");
+  if (is_binary)
+  {
+    i32_buf = (int32_t *)malloc(num_elems * sizeof(int32_t));
+    for (i = 0; i < num_elems; i++)
+    {
+      eid = elem_indices[i];
+      i32_buf[i] = elem[eid].mat;
+    }
+    write_base64_data_array(fp, i32_buf, num_elems * sizeof(int32_t));
+    free(i32_buf);
+  }
+  else
+  {
+    for (i = 0; i < num_elems; i++)
+    {
+      eid = elem_indices[i];
+      fprintf(fp, "\n          %d", elem[eid].mat);
+    }
+    fprintf(fp, "\n        ");
+  }
+  fprintf(fp, "</DataArray>\n");
+  fprintf(fp, "      </CellData>\n");
+
+  /* 3. Point Coordinates */
+  fprintf(fp, "      <Points>\n");
+  fprintf(fp, "        <DataArray type=\"Float64\" Name=\"Points\" NumberOfComponents=\"3\" format=\"%s\">",
+          is_binary ? "binary" : "ascii");
+  if (is_binary)
+  {
+    f64_buf = (double *)malloc(num_pts * 3 * sizeof(double));
+    for (i = 0; i < num_pts; i++)
+    {
+      nid = pts_nodenr[i];
+      f64_buf[i * 3 + 0] = node[nid].nx;
+      f64_buf[i * 3 + 1] = node[nid].ny;
+      f64_buf[i * 3 + 2] = node[nid].nz;
+    }
+    write_base64_data_array(fp, f64_buf, num_pts * 3 * sizeof(double));
+    free(f64_buf);
+  }
+  else
+  {
+    for (i = 0; i < num_pts; i++)
+    {
+      nid = pts_nodenr[i];
+      fprintf(fp, "\n          %.9e %.9e %.9e", node[nid].nx, node[nid].ny, node[nid].nz);
+    }
+    fprintf(fp, "\n        ");
+  }
+  fprintf(fp, "</DataArray>\n");
+  fprintf(fp, "      </Points>\n");
+
+  /* 4. Cells (Topology, Offsets, Types) */
+  fprintf(fp, "      <Cells>\n");
 
-  fprintf(fp, "        <DataArray type=\"Int32\" Name=\"Material_ID\" format=\"ascii\">\n");
+  /* Count total connectivity points */
+  total_conn = 0;
   for (i = 0; i < num_elems; i++)
   {
     eid = elem_indices[i];
-    fprintf(fp, "          %d\n", elem[eid].mat);
+    get_vtk_cell_type(elem[eid].type, &n_nodes);
+    total_conn += n_nodes;
   }
-  fprintf(fp, "        </DataArray>\n");
-  fprintf(fp, "      </CellData>\n");
 
-  fprintf(fp, "    </Piece>\n");
-  fprintf(fp, "  </UnstructuredGrid>\n");
-  fprintf(fp, "</VTKFile>\n");
+  /* Connectivity */
+  fprintf(fp, "        <DataArray type=\"Int32\" Name=\"connectivity\" format=\"%s\">",
+          is_binary ? "binary" : "ascii");
+  if (is_binary)
+  {
+    int c_idx = 0;
+    i32_buf = (int32_t *)malloc(total_conn * sizeof(int32_t));
+    for (i = 0; i < num_elems; i++)
+    {
+      eid = elem_indices[i];
+      vtk_type = get_vtk_cell_type(elem[eid].type, &n_nodes);
+      for (j = 0; j < n_nodes; j++)
+      {
+        int map_idx = j;
+        if (elem[eid].type == 4) map_idx = vtk_map_hex20[j];
+        else if (elem[eid].type == 5) map_idx = vtk_map_penta15[j];
+
+        nid = elem[eid].nod[map_idx];
+        i32_buf[c_idx++] = (nid >= 0 && nid <= anz->nmax) ? (int32_t)nodeMap[nid] : 0;
+      }
+    }
+    write_base64_data_array(fp, i32_buf, total_conn * sizeof(int32_t));
+    free(i32_buf);
+  }
+  else
+  {
+    for (i = 0; i < num_elems; i++)
+    {
+      eid = elem_indices[i];
+      vtk_type = get_vtk_cell_type(elem[eid].type, &n_nodes);
+      fprintf(fp, "\n         ");
+      for (j = 0; j < n_nodes; j++)
+      {
+        int map_idx = j;
+        if (elem[eid].type == 4) map_idx = vtk_map_hex20[j];
+        else if (elem[eid].type == 5) map_idx = vtk_map_penta15[j];
+
+        nid = elem[eid].nod[map_idx];
+        fprintf(fp, " %d", (nid >= 0 && nid <= anz->nmax) ? nodeMap[nid] : 0);
+      }
+    }
+    fprintf(fp, "\n        ");
+  }
+  fprintf(fp, "</DataArray>\n");
+
+  /* Offsets */
+  fprintf(fp, "        <DataArray type=\"Int32\" Name=\"offsets\" format=\"%s\">",
+          is_binary ? "binary" : "ascii");
+  if (is_binary)
+  {
+    offset = 0;
+    i32_buf = (int32_t *)malloc(num_elems * sizeof(int32_t));
+    for (i = 0; i < num_elems; i++)
+    {
+      eid = elem_indices[i];
+      get_vtk_cell_type(elem[eid].type, &n_nodes);
+      offset += n_nodes;
+      i32_buf[i] = (int32_t)offset;
+    }
+    write_base64_data_array(fp, i32_buf, num_elems * sizeof(int32_t));
+    free(i32_buf);
+  }
+  else
+  {
+    offset = 0;
+    for (i = 0; i < num_elems; i++)
+    {
+      eid = elem_indices[i];
+      get_vtk_cell_type(elem[eid].type, &n_nodes);
+      offset += n_nodes;
+      fprintf(fp, "\n          %ld", offset);
+    }
+    fprintf(fp, "\n        ");
+  }
+  fprintf(fp, "</DataArray>\n");
+
+  /* Types */
+  fprintf(fp, "        <DataArray type=\"UInt8\" Name=\"types\" format=\"%s\">",
+          is_binary ? "binary" : "ascii");
+  if (is_binary)
+  {
+    u8_buf = (uint8_t *)malloc(num_elems * sizeof(uint8_t));
+    for (i = 0; i < num_elems; i++)
+    {
+      eid = elem_indices[i];
+      u8_buf[i] = (uint8_t)get_vtk_cell_type(elem[eid].type, &n_nodes);
+    }
+    write_base64_data_array(fp, u8_buf, num_elems * sizeof(uint8_t));
+    free(u8_buf);
+  }
+  else
+  {
+    for (i = 0; i < num_elems; i++)
+    {
+      eid = elem_indices[i];
+      vtk_type = get_vtk_cell_type(elem[eid].type, &n_nodes);
+      fprintf(fp, "\n          %d", vtk_type);
+    }
+    fprintf(fp, "\n        ");
+  }
+  fprintf(fp, "</DataArray>\n");
+  fprintf(fp, "      </Cells>\n");
+
+    fprintf(fp, "    </Piece>\n");
+    fprintf(fp, "  </UnstructuredGrid>\n");
+    fprintf(fp, "</VTKFile>\n");
 
   fclose(fp);
-  printf("  -> Written: %s (%d nodes, %d cells)\n", filename, num_pts, num_elems);
+  printf("  -> Written: %s (%d nodes, %d cells, %s)\n", filename, num_pts, num_elems, is_binary ? "binary" : "ascii");
   return 1;
 }
 
@@ -387,8 +713,8 @@ typedef struct {
   int    lcs[64];
 } PhysStep;
 
-/* Main entry point for send <set> vtu/vtk [all] */
-int write2vtu(char *setname, char *format, int strings, char **string, Summen *anz,
+/* Main entry point for send <set> vtu [all] [ascii] */
+int write2vtu(char *setname, int strings, char **string, Summen *anz,
               Nodes *node, Faces *face, Elements *elem, Sets *set, Datasets *lcase)
 {
   int setNr, i, j, k, eid, nid, n_nodes, vtk_type;
@@ -397,7 +723,7 @@ int write2vtu(char *setname, char *format, int strings, char **string, Summen *a
   int *pts_nodenr = NULL;
   int *elem_indices = NULL;
   int export_all_steps = 0;
-  char ext[8] = "vtu";
+  int is_binary = 1; /* Default to high-performance Base64 binary */
   char vtu_filename[MAX_LINE_LENGTH];
   char pvd_filename[MAX_LINE_LENGTH];
   FILE *pvd_fp = NULL;
@@ -406,9 +732,6 @@ int write2vtu(char *setname, char *format, int strings, char **string, Summen *a
   int num_psteps = 0;
   int active_pstep = 0;
 
-  if (format != NULL && compare(format, "vtk", 3) == 3) strcpy(ext, "vtk");
-  else if (format != NULL && compare(format, "vtu", 3) == 3) strcpy(ext, "vtu");
-
   if (anz->n == 0 || anz->e == 0)
   {
     printf("ERROR: No mesh available to export.\n");
@@ -422,17 +745,19 @@ int write2vtu(char *setname, char *format, int strings, char **string, Summen *a
     return 0;
   }
 
-  /* Check if \"all\" steps requested */
-  if (strings > 0 && string != NULL && string[0] != NULL)
+  /* Parse options (all, ascii/txt) */
+  for (i = 0; i < strings; i++)
   {
-    if (compare(string[0], "all", 3) == 3)
+    if (string[i] != NULL)
     {
-      export_all_steps = 1;
+      if (compare(string[i], "all", 3) == 3) export_all_steps = 1;
+      else if (compare(string[i], "asc", 3) == 3 || compare(string[i], "txt", 3) == 3) is_binary = 0;
     }
   }
 
-  printf("\n=== Exporting to VTK XML Unstructured Grid (.%s) ===\n", ext);
-  printf("Set: %s (mode: %s)\n", setname, export_all_steps ? "all steps (.pvd)" : "active step (.vtu)");
+  printf("\n=== Exporting to VTK XML Unstructured Grid (.vtu) ===\n");
+  printf("Set: %s (mode: %s, encoding: %s)\n",
+         setname, export_all_steps ? "all steps (.pvd)" : "active step (.vtu)", is_binary ? "binary (Base64)" : "ascii");
 
   /* 1. Identify Elements in Set */
   num_elems = set[setNr].anz_e;
@@ -547,9 +872,10 @@ int write2vtu(char *setname, char *format, int strings, char **string, Summen *a
     /* Write individual physical step files */
     for (i = 0; i < num_psteps; i++)
     {
-      sprintf(vtu_filename, "%s_step_%03d.%s", setname, i + 1, ext);
+      sprintf(vtu_filename, "%s_step_%03d.vtu", setname, i + 1);
       write_single_vtu_file(vtu_filename, psteps[i].num_lcs, psteps[i].lcs,
                             psteps[i].value, psteps[i].step_number, psteps[i].analysis_type,
+                            is_binary,
                             anz, node, elem, set, setNr, lcase,
                             num_pts, pts_nodenr, nodeMap, num_elems, elem_indices);
 
@@ -586,11 +912,12 @@ int write2vtu(char *setname, char *format, int strings, char **string, Summen *a
   else
   {
     /* Single active step / mesh */
-    sprintf(vtu_filename, "%s.%s", setname, ext);
+    sprintf(vtu_filename, "%s.vtu", setname);
     if (num_psteps > 0)
     {
       write_single_vtu_file(vtu_filename, psteps[active_pstep].num_lcs, psteps[active_pstep].lcs,
                             psteps[active_pstep].value, psteps[active_pstep].step_number, psteps[active_pstep].analysis_type,
+                            is_binary,
                             anz, node, elem, set, setNr, lcase,
                             num_pts, pts_nodenr, nodeMap, num_elems, elem_indices);
     }
@@ -598,6 +925,7 @@ int write2vtu(char *setname, char *format, int strings, char **string, Summen *a
     {
       write_single_vtu_file(vtu_filename, 0, NULL,
                             0.0, 0, 0,
+                            is_binary,
                             anz, node, elem, set, setNr, lcase,
                             num_pts, pts_nodenr, nodeMap, num_elems, elem_indices);
     }
-- 
2.39.3 (Apple Git-146)

